src/devices/somfy_iohc.c

```diff
@@ -91,60 +91,123 @@ Example packets:
+    msg.dst_addr = (b[2] << 16) | (b[3] << 8) | b[4];
+    msg.src_addr = (b[5] << 16) | (b[6] << 8) | b[7];
+    msg.cmd_id = b[8];
+
+    unsigned int data_length = msg.frame_length - 8;
+    if (msg.protocol_mode == 0 || data_length < 8) {
+        bitrow_snprint(&b[9], data_length * 8, msg.data, sizeof msg.data);
+    } else {
+        data_length -= 8;
+        bitrow_snprint(&b[9], data_length * 8, msg.data, sizeof msg.data);
+        msg.seq_nr = (b[9 + data_length] << 8) | b[9 + data_length + 1];
+        bitrow_snprint(&b[9 + data_length + 2], 6 * 8, msg.mac, sizeof msg.mac);
+    }
+
+    msg.crc = (b[len - 2] << 8) | b[len - 1];
 
     // calculate and verify checksum
     if (crc16lsb(b, len, 0x8408, 0x0000) != 0) // unreflected poly 0x1021
         return DECODE_FAIL_MIC;
-    decoder_logf_bitrow(decoder, 2, __func__, b, len * 8, "offset %u, num_bits %u, len %d, msg_len %d", offset, num_bits, len, msg_len);
 
-    int msg_type = (b[0]);
-    int dst_id   = ((unsigned)b[4] << 24) | (b[3] << 16) | (b[2] << 8) | (b[1]); // assume Little-Endian
-    int src_id   = ((unsigned)b[8] << 24) | (b[7] << 16) | (b[6] << 8) | (b[5]); // assume Little-Endian
-    int counter  = (b[len - 10] << 8) | (b[len - 9]);
-    char msg_str[15 * 2 + 1];
-    bitrow_snprint(&b[9], msg_len * 8, msg_str, 15 * 2 + 1);
-    char mac_str[13];
-    bitrow_snprint(&b[len - 8], 6 * 8, mac_str, 13);
+    decoder_logf_bitrow(decoder, 2, __func__, b, len * 8, "offset %u, num_bits %u, len %d, msg_len %d", offset, num_bits, len, msg.frame_length);
 
     /* clang-format off */
     data_t *data = data_make(
             "model",            "",                 DATA_STRING, "Somfy-IOHC",
-            "id",               "",                 DATA_FORMAT, "%08x", DATA_INT, src_id,
-            "dst_id",           "Dest ID",          DATA_FORMAT, "%08x", DATA_INT, dst_id,
-            "msg_type",         "Msg type",         DATA_FORMAT, "%02x", DATA_INT, msg_type,
-            "msg",              "Message",          DATA_STRING, msg_str,
-            "counter",          "Counter",          DATA_INT,    counter,
-            "mac",              "MAC",              DATA_STRING, mac_str,
+            "id",               "Source",           DATA_FORMAT, "%06x", DATA_INT, msg.src_addr,
+            "dst_id",           "Target",           DATA_FORMAT, "%06x", DATA_INT, msg.dst_addr,
+            "msg_type",         "Command",          DATA_FORMAT, "%02x", DATA_INT, msg.cmd_id,
+            "msg",              "Message",          DATA_STRING, msg.data,
             "mic",              "Integrity",        DATA_STRING, "CRC",
+            "mode",             "Mode",             DATA_STRING, msg.protocol_mode ? "One-way" : "Two-way",
+            "version",          "Version",          DATA_INT, msg.protocol_version,
+            NULL);
+
+    if (msg.protocol_mode == 1)
+        data = data_append(data,
+            "counter",          "Counter",          DATA_INT,    msg.seq_nr,
```

Comment:
You can also use `DATA_COND` here an put in the data_make above.
E.g. 
`"counter",          "Counter",          DATA_COND, msg.protocol_mode == 1, DATA_INT,    msg.seq_nr,`